projects/electron-ngrx/src/lib/window-communication.service.ts

```diff
@@ -1,59 +1,112 @@
 import {Injectable} from '@angular/core';
 import {fromEvent, Observable} from 'rxjs';
-import {filter, map, withLatestFrom} from 'rxjs/operators';
+import {filter, map, pluck, share, take, withLatestFrom} from 'rxjs/operators';
 import {NavigationEnd, Router} from '@angular/router';
+import {ipcRenderer, remote} from 'electron';
+
+interface WithSender {
+  data: any;
+  senderId: number;
+  massageId: number;
+}
+
+export interface Massage<T> {
+  reply: (...arg: any) => void;
+  data: T;
+}
+
+export const enum chancels {
+  replay = 'replay',
+  route = 'route',
+  id = 'id',
+  parent = 'parent'
+}
 
 @Injectable()
 export class WindowCommunicationService {
-  ipcRenderer;
-  remote;
+  ipcRenderer: typeof ipcRenderer;
+  remote: typeof remote;
+  private readonly myId: number;
+  private reply: Observable<any>;
 
   constructor(private router: Router) {
     const electron = (window as any).require('electron');
     this.ipcRenderer = electron.ipcRenderer;
     this.remote = electron.remote;
+    this.myId = electron.remote.getCurrentWindow().id;
+    this.reply = fromEvent(this.ipcRenderer, chancels.replay).pipe(map(this.fromElectronMassage), share());
   }
 
-  listenToChild() {
-    return fromEvent(this.ipcRenderer, 'send-to-parent')
-      .pipe(map(this.stripData));
+  listenToChild<T>(): Observable<Massage<T>> {
+    return this.listenIpcRenderer<T>(chancels.parent);
   }
 
-  listenToId() {
-    return fromEvent(this.ipcRenderer, 'send-to-id')
-      .pipe(map(this.stripData));
+  listenToId<T>(): Observable<Massage<T>> {
+    return this.listenIpcRenderer<T>(chancels.id);
   }
 
-  listenToRoute() {
-    return fromEvent(this.remote.ipcMain, 'send-to-route').pipe(
-      map(this.stripData),
-      withLatestFrom(this.getWindowRoute(), (data, url) => ({data, url})),
-      filter(({url, data}) => url === data.url),
-      map(({data}) => data.payload)
+  listenToRoute<T>(): Observable<{ data: T }> {
+    return fromEvent(this.remote.ipcMain, chancels.route).pipe(
```

Comment:
use this.listenIpcRenderer like all the other listen methods do

Reply:
but it's listening to ipc main